secrets/provider_registry.go

```diff
@@ -0,0 +1,131 @@
+package secrets
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/buildkite/agent/v3/bootstrap/shell"
+	"github.com/puzpuzpuz/xsync"
+)
+
+type ProviderRegistry struct {
+	config ProviderRegistryConfig
+	// We have the two maps here because we only want to initialise a provider if a secret using that provider is used
+	// We shouldn't boot up any secrets providers that won't get used
+	candidates *xsync.MapOf[string, providerCandidate] // Candidates technically doesn't need to be a sync map as it's never altered after initialization, but i've made it one just for symmetry
+	providers  *xsync.MapOf[string, Provider]
+}
+
+type ProviderRegistryConfig struct {
+	Shell *shell.Shell
+}
+
+// NewProviderRegistryFromJSON takes a JSON string representing a slice of secrets.RawProvider, and returns a ProviderRegistry,
+// ready to be used to fetch secrets.
+func NewProviderRegistryFromJSON(config ProviderRegistryConfig, jsonIn string) (*ProviderRegistry, error) {
+	var rawProviders []providerCandidate
+	err := json.Unmarshal([]byte(jsonIn), &rawProviders)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshalling secret providers: %w", err)
+	}
+
+	candidates := xsync.NewMapOf[providerCandidate]()
+	for _, provider := range rawProviders {
+		if _, ok := candidates.Load(provider.ID); ok {
+			return nil, fmt.Errorf("duplicate provider ID: %s. Provider IDs must be unique", provider.ID)
+		}
+
+		candidates.Store(provider.ID, provider)
+	}
+
+	return &ProviderRegistry{
+		config:     config,
+		candidates: candidates,
+		providers:  xsync.NewMapOf[Provider](),
+	}, nil
+}
+
+func (pr *ProviderRegistry) FetchAll(configs []SecretConfig) ([]Secret, []error) {
+	secrets := make([]Secret, 0, len(configs))
+	secretsCh := make(chan Secret)
+
+	errors := make([]error, 0, len(configs))
+	errorsCh := make(chan error)
+
+	var wg sync.WaitGroup
+	for _, c := range configs {
+		wg.Add(1)
+		go func(config SecretConfig) {
+			defer wg.Done()
+
+			secret, err := pr.Fetch(config)
+			if err != nil {
+
+				errorsCh <- err
+				return
+			}
+			secretsCh <- secret
+		}(c)
+	}
+
+	go func() {
+		for err := range errorsCh {
+			errors = append(errors, err)
+		}
+	}()
+
+	go func() {
+		for secret := range secretsCh {
+			secrets = append(secrets, secret)
+		}
+	}()
```

Comment:
maybe this is getting too goroutiney. do we really need to fetch things in parallel? we can, so we might as well, was my justification, but i think the reality is that the number of secrets in each pipeline is gonna be pretty small, and they could probably be serialised.

Reply:
If you keep the parallelisation, might want to at least cap the number of worker goroutines at runtime.NumCPU or similar.

Also is it worth returning each error, or should it treat the whole operation as failed if any of them fail?

Also while you've been out, I've been busy sprinkling context.Context args through the codebase - can requests be cancelled early through use of a context somehow?

Reply:
> If you keep the parallelisation, might want to at least cap the number of worker goroutines at runtime.NumCPU or similar

what would the benefit of limiting concurrency be? given that calls to secret managers are going to be IO bound, i'm happy to let greenthreading/goroutine magic handle throttling our concurrency.

> Is it worth returning each error, or should it treat the whole operation as failed if any of them fail?

I think it's probably most useful to do both - ie, any `len(errors) > 0` should be treated as a failure, but it's also nice to be able to tell the user exactly which secret failed to be fetched, and for what reason.

> can requests be cancelled early through use of a context somehow?

yep, i think it's totally reasonable to change the `Provider.Fetch` interface to accept a context. i'll make that change now :)

Reply:
> what would the benefit of limiting concurrency be? given that calls to secret managers are going to be IO bound, i'm happy to let greenthreading/goroutine magic handle throttling our concurrency.

That's the happy path - what if they each open a connection simultaneously, exhausting TCP source ports? Or there's a config error?